### PR TITLE
API 11228 Datamart Immunization Dose Quantity

### PR DESCRIPTION
**Pull request by that-guy-is-it**

### CODE REVIEW

File: data-query/src/main/java/gov/va/api/health/dataquery/service/controller/immunization/DatamartImmunization.java [hunk 1]
```diff
@@ -1,6 +1,7 @@
 package gov.va.api.health.dataquery.service.controller.immunization;
 
 import com.fasterxml.jackson.annotation.JsonProperty;
+import gov.va.api.health.r4.api.datatypes.Quantity;
```
Comment (jblefkove-va) [inline data-query/src/main/java/gov/va/api/health/dataquery/service/controller/immunization/DatamartImmunization.java:4]: This needs to be a Datamart quantity, not FHIR; take a look at DatamartObservation

File: data-query/src/main/java/gov/va/api/health/dataquery/service/controller/immunization/DatamartImmunization.java [hunk 2]
```diff
@@ -36,6 +37,8 @@
 
   private boolean primarySource;
 
+  private Optional<Quantity> doseQuantity;
```
Comment (jblefkove-va) [inline data-query/src/main/java/gov/va/api/health/dataquery/service/controller/immunization/DatamartImmunization.java:40]: will also need lazy getter

File: data-query/src/main/java/gov/va/api/health/dataquery/service/controller/immunization/DatamartImmunization.java [hunk 3]
```diff
@@ -121,6 +131,29 @@ private void setEtlDate(String unused) {
...
+  public static final class Quantity implements AsList<Quantity> {
+    private Double value;
+
+    private String unit;
+
+    private String system;
+
+    private String code;
+
+    void setUnitCode(String code) {
```
Comment (jblefkove-va) [inline data-query/src/main/java/gov/va/api/health/dataquery/service/controller/immunization/DatamartImmunization.java:148]: These setters are for backwards compatibility; we don't need them here

**Comment by jblefkove-va** [comment]
@department-of-veterans-affairs/lighthouse-datasaurus

File: data-query/src/main/java/gov/va/api/health/dataquery/service/controller/immunization/DatamartImmunization.java [hunk 4]
```diff
@@ -121,6 +131,21 @@ private void setEtlDate(String unused) {
     data_absent_reason_unsupported
   }
 
+  /** Quantity. */
+  @Data
+  @Builder
+  @NoArgsConstructor(access = AccessLevel.PRIVATE)
+  @AllArgsConstructor(access = AccessLevel.PRIVATE)
+  public static final class Quantity implements AsList<Quantity> {
+    private Double value;
```
Comment (jblefkove-va) [inline data-query/src/main/java/gov/va/api/health/dataquery/service/controller/immunization/DatamartImmunization.java:140]: @nmoisan if Immunization `doseQuantity` is present, are any of these fields guaranteed non-null?

Comment (nmoisan) [reply data-query/src/main/java/gov/va/api/health/dataquery/service/controller/immunization/DatamartImmunization.java:140]: unfortunately no. there are cases where one or the other is still NULL. but we could force it such that we only send _usable_ data in the payload that has `dosage` and/or `doseUnit` populated. thoughts?

Comment (jblefkove-va) [reply data-query/src/main/java/gov/va/api/health/dataquery/service/controller/immunization/DatamartImmunization.java:140]: Do we do that kind of filtering elsewhere in datamart?

All of the quantity fields in the fhir spec are nullable, so it's not necessarily a problem. But it does indicate that everything should be `Optional<>` here

Comment (nmoisan) [reply data-query/src/main/java/gov/va/api/health/dataquery/service/controller/immunization/DatamartImmunization.java:140]: We do in some cases - but I think mostly in places where a sub-property is `required` if the optional parent is to be sent off. E.G. `Observation.Component.Code`

Comment (nmoisan) [reply data-query/src/main/java/gov/va/api/health/dataquery/service/controller/immunization/DatamartImmunization.java:140]: I'm just wondering how useful `units` is without `dosage`. But yeah Optional is probably safe

File: data-query/src/main/java/gov/va/api/health/dataquery/service/controller/immunization/DatamartImmunization.java [hunk 5]
```diff
@@ -121,6 +131,21 @@ private void setEtlDate(String unused) {
     data_absent_reason_unsupported
   }
 
+  /** Quantity. */
+  @Data
+  @Builder
+  @NoArgsConstructor(access = AccessLevel.PRIVATE)
+  @AllArgsConstructor(access = AccessLevel.PRIVATE)
+  public static final class Quantity implements AsList<Quantity> {
+    private Optional<Double> value;
```
Comment (jblefkove-va) [inline data-query/src/main/java/gov/va/api/health/dataquery/service/controller/immunization/DatamartImmunization.java:140]: Now that these are optional they also need lazy getters...

**Comment by jblefkove-va** [comment]
Update `DatamartImmunizationTest` as well

**Review by jblefkove-va — APPROVED** [review]
:shipit:

**Pull request merged**